fix(replication): prevent dropped conns during replication

joshuef opened the pull request:
We were seeing a lot of OutboundFialure and closed connections leading to dropped messages during data replication. This was turning up as UnexpectedEof errors when a connection was closed midway through send eg.

Seems like, due to _NOT_ responding to a request (using the protocol wrongly perhaps?) we were causing awaiting connections to error out as the receive stream was dropped unepectedly, causing a knock on to anything else in flight on that same connection.

## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 09 Jun 23 02:32 UTC
This pull request fixes dropped messages during data replication due to closed connections from outbound failures. The change includes adding a response to the ReplicateCmd to prevent connection failures and errors in awaiting connections.
<!-- reviewpad:summarize:end -->

bochaco commented on src/prefix_map/mod.rs:
```diff
@@ -214,6 +229,7 @@ impl NetworkPrefixMap {
     /// Get the entry at the prefix that matches `name`. In case of multiple matches, returns the
     /// one with the longest prefix. If there are no prefixes matching the given `name`, return
     /// a prefix matching the opposite to 1st bit of `name`. If the map is empty, return None.
+    #[allow(unused)]
```
Shall we just remove it if we are not using it?

maqi commented on src/prefix_map/mod.rs:
```diff
@@ -68,13 +68,28 @@ impl NetworkPrefixMap {
 
+    /// Returns the known section that is closest to the given name, regardless of whether `name`
+    /// belongs in that section or not. If there are no close matches, return a SAP from an opposite prefix.
+    pub(crate) fn closest_or_opposite(
```
not sure what's the point of `opposite`
the function says `closest` , which means as long as we know a section, this function shall return with a result.
And, in case of `no close matches`, it means we don't know a section, which won't have a result for `opposite` either?

Yoga07 replied:
This is for scenarios where there is just one branch of sections in the PrefixMap(happens mostly at the client). Eg: The map would contain only 10, 11 and so on where prefixes start with just 1. So a `closest` match for a dst name `00111.....` would only return `None`. So as a failsafe, we need to return at least one prefix from the opposite section.

maqi replied:
then what if you have 01, 11, and `closest match` for `00111` ? returns `01` ? 
If as you explained, I think the name of the function is bit misleading, `closest` shall only mean Xor min distance among the sections to the target. Which, shall return a section as long as knows a section.

bochaco commented on src/routing/core/msg_handling/agreement.rs:
```diff
@@ -227,20 +227,38 @@ impl Core {
         }
 
-        self.update_state(signed_section_auth, snapshot).await
+        info!("PREFIXES WE KNOW ABOUT: {:?}", self.network);
```
I guess this msg shall be lower case

Yoga07 commented on src/prefix_map/mod.rs:
```diff
@@ -255,13 +271,12 @@ impl NetworkPrefixMap {
             .sum();
-        let known = our.elder_count() + network_elders_count;
-        let total = known as f64 / network_fraction;
+        let total = network_elders_count as f64 / network_fraction;
```
Nop, `networks_elders_count` now contains elders of our own section.

Yoga07 commented on src/routing/core/mod.rs:
```diff
@@ -142,11 +141,7 @@ impl Core {
-        new_section_auth: SectionAuth<SectionAuthorityProvider>,
-        old: StateSnapshot,
-    ) -> Result<Vec<Command>> {
+    pub(crate) async fn update_state(&mut self, old: StateSnapshot) -> Result<Vec<Command>> {
```
We only use the old state to make comparisons and take decisions on what to do with the churn/split :)

joshuef committed:
fix(replication): prevent dropped conns during replication

We were seeing a lot of OutboundFialure and closed connections leading
to dropped messages during data replication. This was turning up as
UnexpectedEof errors when a connection was closed midway through send
eg.

Seems like, due to _NOT_ responding to a request (using the protocol
wrongly perhaps?) we were causing awaiting connections to error out
as the receive stream was dropped unepectedly, causing a knock on to
anything else in flight on that same connection.
